src/MinimumNodeService.cpp

```diff
@@ -383,104 +384,43 @@ Processed MinimumNodeService::handleMessage(unsigned int opc, CANFrame *msg)
-          }          
-          return PROCESSED;
-          
-        case MODE_SETUP:
-          if (nn != 0)   // Not for this node
-          {
-            return PROCESSED;
-          } 
-          else
-          {
-            controller->sendGRSP(OPC_MODE, getServiceID(), CMDERR_INV_CMD);            
-          }
-          return PROCESSED;
-          
-        case MODE_NORMAL:
-          switch (newMode)
-          {
-          case MODE_SETUP:
+          case 0x00:
+          // Request Setup
             controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
             initSetupFromNormal();
-            break;
-          
-          case MODE_LEARN:
-            //DEBUG_SERIAL << F("> MODE -- Learn") << endl;
-            controller->setLearnMode(newMode);
-            instantMode = newMode;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
-            break;
-          
-          case MODE_NHEARTB:          
-            //DEBUG_SERIAL << F("> MODE -- request no Heartbeats") << endl;
-            noHeartbeat = true;
-            instantMode = newMode;
-            break;
-          }
-          return PROCESSED;
-          
-        case MODE_LEARN:
-          if (newMode == MODE_SETUP)
-          {
-            controller->sendGRSP(OPC_MODE, getServiceID(), CMDERR_INV_CMD); 
-          }
-          if (newMode == MODE_NORMAL)
-          {
-            controller->setLearnMode(newMode);
-            if (noHeartbeat == true)
-            {
-              instantMode = MODE_NHEARTB;
-            }
-            else
-            {
-              instantMode = newMode;
-            }
-          } 
-          return PROCESSED;
-          
-        case MODE_NHEARTB:
-          if (newMode == MODE_SETUP)
-          {
-            noHeartbeat = false;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
-            initSetupFromNormal();
-          }
-          if (newMode == MODE_NORMAL)
-          {
+            return PROCESSED;
+            
+          case 0x0C:
+          // Turn on Heartbeat
             noHeartbeat = false;
-            instantMode = newMode;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
+            module_config->setHeartbeat(!noHeartbeat);
+            return PROCESSED;
+            
+          case 0x0D:
+          // Turn off Heartbeat
+            noHeartbeat = true;
+            module_config->setHeartbeat(!noHeartbeat);
+            return PROCESSED;
           }
-          if (newMode == MODE_LEARN)
-          {
-            controller->setLearnMode(newMode);
-            instantMode = newMode;
-            controller->sendGRSP(OPC_MODE, getServiceID(), GRSP_OK);
-          } 
+          // if in MODE_NORMAL but none of the above commands, let another service see message
+          break;  
+        }
+        else
+        {
           return PROCESSED;
         }
       }
-      
+      else
+      {
       return PROCESSED;
```

Comment:
If you want to keep it, please indent this return statement correctly.

Reply:
I though it was. the else refers to the "if (instantMode == MODE_NORMAL)" statement at line 390

Reply:
Sorry, getting confused about which else statement you are referring to.  Are you, in fact, saying that the statement at line 415 is superfluous? I think that is only the case if I get rid of both else statements at lines 415 and 420 and simple end with "return PROCESSED;"  Do you agree?

Reply:
No It is just that line 422 should be indented relative to the curly bracket on the line above.

Reply:
I sent you a separate email on the "else" style.

The issue about letting other services see MODE commands in other modes than Normal is something we should discuss separately. And changing this behaviour can wait until we see a need for it.

Reply:
Seen the email, which makes a lot of sense and makes things much easier to read.
I agree that we can discuss passing on of the MODE commands.  I also agree that we should wait for a proven need before adding the capability.